test/Microsoft.ML.Tests/CollectionDataSourceTests.cs

```diff
@@ -205,5 +207,314 @@ public class IrisPrediction
+            {
+                var xvalue = field.GetValue(x);
+                var yvalue = field.GetValue(y);
+                if (field.FieldType.IsArray)
+                {
+                    if (!CompareArrayValues(xvalue as Array, yvalue as Array))
+                        return false;
+                }
+                else
+                {
+                    if (!CompareObjectValues(xvalue, yvalue, field.FieldType))
+                        return false;
+                }
+            }
+            return true;
+        }
+
+        public bool CompareArrayValues(Array x, Array y)
+        {
+            if (x == null && y == null) return true;
+            if ((x == null && y != null) || (y == null && x != null))
+                return false;
+            if (x.Length != y.Length)
+                return false;
+            for (int i = 0; i < x.Length; i++)
+                if (!CompareObjectValues(x.GetValue(i), y.GetValue(i), x.GetType().GetElementType()))
+                    return false;
+            return true;
+        }
+
+        public class ClassWithConstField
+        {
+            public const string ConstString = "N";
+            public string fString;
+            public const int ConstInt = 100;
+            public int fInt;
+        }
+
+        [Fact]
+        public void BackAndForthConversionWithBasicTypes()
+        {
+            var data = new List<ConversionSimpleClass>()
+            {
+                new ConversionSimpleClass(){ fInt=int.MaxValue-1, fuInt=uint.MaxValue-1, fBool=true, fsByte=sbyte.MaxValue-1, fByte = byte.MaxValue-1,
+                    fDouble =double.MaxValue-1, fFloat=float.MaxValue-1, fLong=long.MaxValue-1, fuLong = ulong.MaxValue-1,
+                    fShort =short.MaxValue-1, fuShort = ushort.MaxValue-1, fString=null},
+                new ConversionSimpleClass(){ fInt=int.MaxValue, fuInt=uint.MaxValue, fBool=true, fsByte=sbyte.MaxValue, fByte = byte.MaxValue,
+                    fDouble =double.MaxValue, fFloat=float.MaxValue, fLong=long.MaxValue, fuLong = ulong.MaxValue,
+                   fShort =short.MaxValue, fuShort = ushort.MaxValue, fString="ooh"},
+                new ConversionSimpleClass(){ fInt=int.MinValue+1, fuInt=uint.MinValue+1, fBool=true, fsByte=sbyte.MinValue+1, fByte = byte.MinValue+1,
+                    fDouble =double.MinValue+1, fFloat=float.MinValue+1, fLong=long.MinValue+1, fuLong = ulong.MinValue+1,
+                    fShort =short.MinValue+1, fuShort = ushort.MinValue+1, fString=""},
+                new ConversionSimpleClass(){},
+            };
+
+            var dataNullable = new List<ConversionNullalbeClass>()
+            {
+                new ConversionNullalbeClass(){ fInt=int.MaxValue-1, fuInt=uint.MaxValue-1, fBool=true, fsByte=sbyte.MaxValue-1, fByte = byte.MaxValue-1,
+                    fDouble =double.MaxValue-1, fFloat=float.MaxValue-1, fLong=long.MaxValue-1, fuLong = ulong.MaxValue-1,
```

Comment:
> = [](start = 27, length = 2)

Could you Ctrl-K-D for proper spacing? Or just, generally, invoke magic for proper spacing?